docs/proposals/20240517-separate-yurtmanager-clients.md

```diff
@@ -0,0 +1,274 @@
+yurt-coordinator(controller): used for yurt-coordinator-cert/pod-binding and all sub-controllers of yurt-coordinator
+yurt-app-set(webhook): used for yurt-app-set webhook only
+...
+```
+
+2. How to use different RBAC for different controller/webhooks?
+
+After different RBACs are prepared, we should make sure they are properly used by different controllers. We consider two feasible approaches here to achieve this goal.
+
+1) User impersonation
+
+Kubernets provides a [mechanism](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation) that one user can act as another user through impersonation headers. These let requests manually override the user info a request authenticates as. 
+
+2) Token override
+
+We can override the client config' token with the prepared ServiceAccount token which has been bound to prescribed roles or cluster roles.
+
+Considering the additional complexity of the first approach, we choose the second one to override the token directly when building the client for each controller/webhook.
+
+### Solution Introduction
+
+The whole solution consists of two steps: 
+
+1. Generate RBAC
+
+First, prepare RBAC artifacts including role/clusterrole, rolebindidng/clusterrolebinding and serviceaccount for controllers and webhooks in yurt-manager. The specific steps are as following:
+
+1) Use bash scripts to search through the pkg/yurt-manager folder and find the components which need independent RBAC.
+2) Generate clusterrole and roles for those components one by one with `controller-gen`.
+3) Use bash scripts to complement the corresponding rolebinding/clusterrolebinding and serviceaccount artifacts.
+4) Gather all generated materials into one yaml in charts/yurt-manager/templates/yurt-manager-auto-generated.yaml with `kustomize`.
+
+Those steps above are performed automatically when you run `make manifests` command. Basically, yurt-manager developers don't need to know these implementations underhood.
+
+2. Prepare client
+
+Currently, all controllers use a client from `manager.GetClient()` which is provided by controller runtime framework directly. However the framework [doesn't provide any interface](https://github.com/kubernetes-sigs/controller-runtime/issues/2822) to provide different clients by controllers. Therefore, we have to implement a client wrapper to to construct a new client for every component based on the basic client.
+
+```go 
+func GetClientByControllerNameOrDie(mgr manager.Manager, controllerName string) client.Client {
+	// if controllerName is empty, return the base client of manager
+	if controllerName == "" {
+		return mgr.GetClient()
+	}
+
+	clientStore.lock.Lock()
+	defer clientStore.lock.Unlock()
+
+	if cli, ok := clientStore.clientsByName[controllerName]; ok {
+		return cli
+	}
+
+	// check if controller-specific ServiceAccount exist
+	_, err := getOrCreateServiceAccount(mgr.GetClient(), "kube-system", controllerName)
```

Comment:
don't use the hard code `kube-system` for namespace, please use the parameter `WorkingNamespace`.

Reply:
done